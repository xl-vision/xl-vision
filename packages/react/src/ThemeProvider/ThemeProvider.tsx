import { ThemeContext as StyledThemeContext } from '@xl-vision/styled-engine';
import PropTypes from 'prop-types';
import React from 'react';
import { env } from '@xl-vision/utils';
import createTheme, { BaseTheme } from './createTheme';
import ThemeContext from './ThemeContext';

export type ThemeProviderProps = {
  children: React.ReactNode;
  theme: BaseTheme;
};

const ThemeProvider: React.FunctionComponent<ThemeProviderProps> = (props) => {
  const { children, theme } = props;

  const value = React.useMemo(() => {
    return createTheme(theme);
  }, [theme]);

  return (
    // both self ThemeContext and styled-engine ThemeContext
    <ThemeContext.Provider value={value}>
      <StyledThemeContext.Provider value={value}>{children}</StyledThemeContext.Provider>
    </ThemeContext.Provider>
  );
};

if (env.isDevelopment) {
  ThemeProvider.displayName = 'ThemeProvider';

  ThemeProvider.propTypes = {
    children: PropTypes.node.isRequired,
    theme: PropTypes.object.isRequired,
  };
}

export default ThemeProvider;
