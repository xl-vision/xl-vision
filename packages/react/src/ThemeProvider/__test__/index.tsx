import { render } from '@testing-library/react';
import { useEffect } from 'react';
import { ThemeInput, ThemeProvider, useTheme } from '@xl-vision/react';
import { enUS, zhCN } from '@xl-vision/react/locale';

describe('ThemeProvider', () => {
  it('Test swicth locale', () => {
    const fn = jest.fn<any, Array<any>>();

    const Demo = () => {
      const { locale } = useTheme();

      useEffect(() => {
        fn(locale);
      }, [locale]);

      return <div />;
    };

    const { rerender } = render(
      <ThemeProvider locale={enUS}>
        <Demo />
      </ThemeProvider>,
    );

    expect(fn.mock.calls.length).toBe(1);
    expect(fn.mock.calls[0][0]).toStrictEqual(enUS);

    fn.mockClear();

    rerender(
      <ThemeProvider locale={zhCN}>
        <Demo />
      </ThemeProvider>,
    );

    expect(fn.mock.calls.length).toBe(1);
    expect(fn.mock.calls[0][0]).toStrictEqual(zhCN);
    fn.mockClear();
  });

  it('Test swicth clsPrefix', () => {
    const fn = jest.fn<any, Array<any>>();

    const Demo = () => {
      const { clsPrefix } = useTheme();

      useEffect(() => {
        fn(clsPrefix);
      }, [clsPrefix]);

      return <div />;
    };

    const { rerender } = render(
      <ThemeProvider clsPrefix='a'>
        <Demo />
      </ThemeProvider>,
    );

    expect(fn.mock.calls.length).toBe(1);
    expect(fn.mock.calls[0][0]).toBe('a');

    fn.mockClear();

    rerender(
      <ThemeProvider clsPrefix='b'>
        <Demo />
      </ThemeProvider>,
    );

    expect(fn.mock.calls.length).toBe(1);
    expect(fn.mock.calls[0][0]).toBe('b');
    fn.mockClear();
  });

  it('Test swicth size', () => {
    const fn = jest.fn<any, Array<any>>();

    const Demo = () => {
      const { sizeVariant } = useTheme();

      useEffect(() => {
        fn(sizeVariant);
      }, [sizeVariant]);

      return <div />;
    };

    const { rerender } = render(
      <ThemeProvider sizeVariant='small'>
        <Demo />
      </ThemeProvider>,
    );

    expect(fn.mock.calls.length).toBe(1);
    expect(fn.mock.calls[0][0]).toBe('small');

    fn.mockClear();

    rerender(
      <ThemeProvider sizeVariant='large'>
        <Demo />
      </ThemeProvider>,
    );

    expect(fn.mock.calls.length).toBe(1);
    expect(fn.mock.calls[0][0]).toBe('large');
    fn.mockClear();
  });

  it('test theme prop override', () => {
    const theme1: ThemeInput = {
      breakpoints: {
        unit: 'px',
      },
    };

    const theme2: ThemeInput = {
      breakpoints: {
        unit: 'rem',
      },
    };

    const Demo1 = () => {
      const theme = useTheme();

      expect(theme.breakpoints.unit).toBe('px');
      return <div />;
    };

    const Demo2 = () => {
      const theme = useTheme();

      expect(theme.breakpoints.unit).toBe('rem');
      return <div />;
    };

    render(
      <ThemeProvider {...theme1}>
        <Demo1 />
        <ThemeProvider {...theme2}>
          <Demo2 />
        </ThemeProvider>
      </ThemeProvider>,
    );

    expect.assertions(2);
  });
});
