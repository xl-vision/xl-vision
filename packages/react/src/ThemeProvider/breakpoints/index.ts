export type Breakpoints = Partial<{
  values: Record<string, number>;
  unit: string;
  column: number;
}>;

const createBreakpoints = (breakpoints: Breakpoints = {}) => {
  const {
    values = {
      xs: 0,
      sm: 576,
      md: 768,
      lg: 992,
      xl: 1200,
      xxl: 1600,
    },
    column = 24,
    unit = 'px',
  } = breakpoints;

  return {
    values,
    unit,
    column,
  };
};

export default createBreakpoints;
