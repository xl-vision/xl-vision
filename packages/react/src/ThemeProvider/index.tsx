import PropTypes from 'prop-types';
import React from 'react';
import { ThemeContext, createTheme, BaseTheme } from '../styles';

export type ThemeProviderProps = {
  children: React.ReactNode;
  theme: BaseTheme;
};

const ThemeProvider: React.FunctionComponent<ThemeProviderProps> = (props) => {
  const { children, theme } = props;

  const value = React.useMemo(() => {
    return createTheme(theme);
  }, [theme]);

  return <ThemeContext.Provider value={value}>{children}</ThemeContext.Provider>;
};

ThemeProvider.propTypes = {
  children: PropTypes.node.isRequired,
  theme: PropTypes.object.isRequired,
};

export default ThemeProvider;
