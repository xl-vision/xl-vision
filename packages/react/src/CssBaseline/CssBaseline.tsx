import { CSSObject } from '@xl-vision/styled-engine';
import React from 'react';
import PropTypes from 'prop-types';
import { isProduction } from '@xl-vision/utils';
import { createGlobalStyles } from '../styles';
import { Theme, useTheme } from '../ThemeProvider';

export type CssBaselineProps = {
  children?: React.ReactNode;
};

export type CssBaselineStyleProps = { overrideStyle?: (theme: Theme) => CSSObject };

const GlobalStyles = createGlobalStyles<CssBaselineStyleProps>(({ theme, styleProps }) => {
  const { overrideStyle } = styleProps;

  const styles = overrideStyle?.(theme);

  return [
    {
      html: {
        WebkitFontSmoothing: 'antialiased', // Antialiasing.
        MozOsxFontSmoothing: 'grayscale', // Antialiasing.
        // Change from `box-sizing: content-box` so that `width`
        // is not affected by `padding` or `border`.
        boxSizing: 'border-box',
        // Fix font resize problem in iOS
        WebkitTextSizeAdjust: '100%',
      },
      '*, *::before, *::after': {
        boxSizing: 'inherit',
      },
      'strong, b': {
        fontWeight: theme.typography.fontWeight.bold,
      },
      body: {
        ...theme.typography.body1.style,
        color: theme.color.text.primary,
        backgroundColor: theme.color.background.default,
        margin: 0,
        '&::backdrop': {
          backgroundColor: theme.color.background.default,
        },
        '@media print': {
          // Save printer ink.
          backgroundColor: '#ffffff',
        },
      },
    },
    styles,
  ];
});

const CssBaseline: React.FunctionComponent<CssBaselineProps> = (props) => {
  const { children } = props;
  const { overrideStyles } = useTheme();

  const overrideStyle = overrideStyles.CssBaseline?.Root;

  return (
    <>
      <GlobalStyles styleProps={{ overrideStyle }} />
      {children}
    </>
  );
};

if (!isProduction) {
  CssBaseline.displayName = 'CssBaseline';
  CssBaseline.propTypes = {
    children: PropTypes.node,
  };
}

export default CssBaseline;
