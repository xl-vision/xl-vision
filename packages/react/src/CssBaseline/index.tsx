import { CSSObject } from '@xl-vision/styled-engine-types';
import React from 'react';
import PropTypes from 'prop-types';
import { env } from '@xl-vision/utils';
import { createGlobalStyles } from '../styles';
import { Theme } from '../ThemeProvider';
import ThemeContext from '../ThemeProvider/ThemeContext';

export type CssBaselineProps = {
  children?: React.ReactNode;
};

export type CssBaselineStyleProps = { overrideStyle?: (theme: Theme) => CSSObject };

const GlobalStyles = createGlobalStyles<CssBaselineStyleProps>(({ theme, styleProps }) => {
  const { overrideStyle } = styleProps;

  const styles = overrideStyle?.(theme);

  return [
    {
      html: {
        WebkitFontSmoothing: 'antialiased', // Antialiasing.
        MozOsxFontSmoothing: 'grayscale', // Antialiasing.
        // Change from `box-sizing: content-box` so that `width`
        // is not affected by `padding` or `border`.
        boxSizing: 'border-box',
        // Fix font resize problem in iOS
        WebkitTextSizeAdjust: '100%',
      },
      '*, *::before, *::after': {
        boxSizing: 'inherit',
      },
      'strong, b': {
        fontWeight: theme.typography.fontWeight.bold,
      },
      body: {
        ...theme.typography.body1,
        color: theme.color.text.primary,
        backgroundColor: theme.color.background.default,
        margin: 0,
        '&::backdrop': {
          backgroundColor: theme.color.background.default,
        },
        '@media print': {
          // Save printer ink.
          backgroundColor: '#ffffff',
        },
      },
    },
    styles,
  ];
});

const CssBaseline: React.FunctionComponent<CssBaselineProps> = (props) => {
  const { children } = props;
  const { overrideStyles } = React.useContext(ThemeContext);

  const overrideStyle = overrideStyles.CssBaseline?.Root;

  return (
    <>
      <GlobalStyles styleProps={{ overrideStyle }} />
      {children}
    </>
  );
};

if (env.isDevelopment) {
  CssBaseline.displayName = 'CssBaseline';
  CssBaseline.propTypes = {
    children: PropTypes.node,
  };
}

export default CssBaseline;
