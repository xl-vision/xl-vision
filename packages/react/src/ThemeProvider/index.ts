import ThemeProvider from './ThemeProvider';

export * from './ThemeProvider';
export * from './createTheme';
export { default as createTheme } from './createTheme';
export { default as ThemeContext } from './ThemeContext';
export default ThemeProvider;
