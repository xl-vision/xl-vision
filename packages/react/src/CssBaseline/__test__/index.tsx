import React from 'react';
import { mount } from 'enzyme';
import CssBaseline from '..';
import ThemeProvider, { BaseTheme } from '../../ThemeProvider';

describe('CssBaseline', () => {
  it.todo('todo here');
  it('basic renders', () => {
    mount(
      <CssBaseline>
        <div />
      </CssBaseline>,
    );

    expect(document.head).toMatchSnapshot();
  });

  it('supports theme overrides', () => {
    const baseTheme: BaseTheme = {
      overrideStyles: {
        CssBaseline: {
          Root(theme) {
            return {
              body: {
                color: theme.color.themes.primary.color,
              },
            };
          },
        },
      },
    };

    mount(
      <ThemeProvider theme={baseTheme}>
        <CssBaseline>
          <div />
        </CssBaseline>
      </ThemeProvider>,
    );

    expect(document.head).toMatchSnapshot();
  });
});
