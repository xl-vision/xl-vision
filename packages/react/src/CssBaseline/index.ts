import CssBaseline from './CssBaseline';

export default CssBaseline;
export * from './CssBaseline';
