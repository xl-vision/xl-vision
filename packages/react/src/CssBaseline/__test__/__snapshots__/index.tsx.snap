// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CssBaseline basic renders 1`] = `<head />`;

exports[`CssBaseline supports theme overrides 1`] = `<head />`;
